### PR TITLE
Calculate cost (for display, not for billing) for OpenAI text-generation models

### PR DESCRIPTION
**Pull request by Rindrics**
## Summary

enables:
- cost calculation of OpenAI text-generation models inside Giselle
  - for displaying preliminary value, not for billing
- telemetry emission for  Langfuse

## Related Issue

- part of #796

## Changes

- add pricing table for gpt models
- implement token-based calculation logic for OpenAI text-generation models
- introduce explicit Langfuse trace

## Testing

- corresponding trace arrived at Langfuse ✅ 
  - image
  - auto-instrumented trace from AISDK also arrives
    - we can distinguish the auto-instrumented trace with tag:
      - image

    - we will deprecate the auto-instrumented trace after #796 is completed

## Other information

We will implement cost calculation for billing in https://github.com/giselles-ai/giselle/issues/865

### CODE REVIEW

File: packages/language-model/src/costs/calculator.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+ * and be defined in their respective provider files.
+ */
+export interface CostCalculator<TUsage extends ModelUsage = TokenUsage> {
+	calculate(model: string, usage: TUsage): Promise<CostResult>;
+}
+
+export class DefaultCostCalculator implements CostCalculator {
+	constructor(private provider: string) {}
+
+	async calculate(): Promise<CostResult> {
```
[the remark this answers is not in the document]
Comment (rindrics) [reply packages/language-model/src/costs/calculator.ts]: fixed: ed9bc828

File: packages/language-model/src/costs/pricing.ts [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+export type TokenBasedPricing = {
+	input: BaseTokenPrice;
+	output: BaseTokenPrice;
+};
+
+export type ModelPrice = {
+	validFrom: string;
+	price: TokenBasedPricing;
+};
+
```
[the remark this answers is not in the document]
Comment (rindrics) [reply packages/language-model/src/costs/pricing.ts:25]: fixed: e76f5953

File: packages/language-model/src/index.ts [hunk 3]
```diff
@@ -25,6 +25,9 @@ import {
 } from "./perplexity";
 export * from "./base";
 export * from "./helper";
+export * from "./costs";
+export * from "./costs/pricing";
+export * from "./costs/model-prices";
```
[the remark this answers is not in the document]
Comment (rindrics) [reply packages/language-model/src/index.ts]: fixed: 81390a5f

File: packages/giselle-engine/src/core/generations/generate-text.ts [hunk 4]
```diff
@@ -59,6 +60,9 @@ export async function generateText(args: {
 	if (!isTextGenerationNode(operationNode)) {
 		throw new Error("Invalid generation type");
 	}
+
+	const langfuse = new Langfuse();
```
[the remark this answers is not in the document]
Comment (rindrics) [reply packages/giselle-engine/src/core/generations/generate-text.ts]: fixed: edfa68a1

File: packages/language-model/src/costs/model-prices.ts [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+	modelId: string,
+	priceTable: ModelPriceTable,
+): ModelPrice {
+	const modelPricing = priceTable[modelId];
+	if (!modelPricing) {
+		throw new Error(`No pricing found for model ${modelId}`);
+	}
+
+	const now = new Date();
+	const validPrices = modelPricing.prices
```
[the remark this answers is not in the document]
Comment (rindrics) [reply packages/language-model/src/costs/model-prices.ts:124]: fixed: ea8d349d

File: packages/giselle-engine/src/core/generations/generate-text.ts [hunk 6]
```diff
@@ -73,6 +77,23 @@ export async function generateText(args: {
...
 
+	const trace = langfuse.trace({
+		sessionId: runningGeneration.id,
+		name: "ai.streamText",
+		metadata: args.telemetry?.metadata,
+	});
+	const span = trace.span({
+		name: "ai.streamText",
+		startTime: new Date(runningGeneration.queuedAt),
+	});
```
[the remark this answers is not in the document]
Comment (rindrics) [reply packages/giselle-engine/src/core/generations/generate-text.ts]: we don't need these comment because Langfuse SDK provides document: https://langfuse.com/docs/sdk/typescript/guide#trace

File: packages/language-model/src/costs/pricing.ts [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+		totalCost: Cost;
+	} | null;
+}
+
+export class TokenBasedCostCalculator implements CostCalculator {
+	constructor(
+		private readonly pricing: Record<string, { prices: ModelPrice[] }>,
+	) {}
+
+	calculateCost(modelId: string, usage: TokenUsage) {
```
Comment (shige) [inline packages/language-model/src/costs/pricing.ts]: @Rindrics 

I'd like you to add test code for `calculateCost`.

File: packages/language-model/src/costs/pricing.ts [hunk 8]
```diff
@@ -0,0 +1,63 @@
+// Model cost in USD
+export type Cost = number;
+
+export function tokensToMegaTokens(tokens: number): number {
+	return tokens / 1_000_000;
```
Comment (shige) [inline packages/language-model/src/costs/pricing.ts:5]: ### Findings and Recommendations on Floating-Point Calculations

The review of the pull request and relevant floating-point calculations reveals the following:

---

### Key Areas Involving Floating-Point Calculations:
1. **Cost Calculations in `packages/language-model/src/costs/pricing.ts`**:
   - **Code**: The `tokensToMegaTokens` function converts token counts to "mega tokens" (1,000,000 tokens).
   ```typescript
   export function tokensToMegaTokens(tokens: number): number {
       return tokens / 1_000_000;
   }
   ```
   - **Risk**: Division operations may introduce precision errors, especially for very large or very small token values. For example, `tokens = 999_999` would result in a fractional value when divided.
   - **Mitigation**: Consider rounding results to a fixed precision if exact values are required, e.g., `toFixed(6)`.

2. **Cost Calculation in `calculateTokenCost`**:
   - **Code**: This function computes costs using floating-point arithmetic:
   ```typescript
   const inputCost = tokensToMegaTokens(usage.input) * pricing.input.costPerMegaToken;
   const outputCost = tokensToMegaTokens(usage.output) * pricing.output.costPerMegaToken;
   ```
   - **Risk**: Multiplication of floating-point numbers can accumulate errors, especially when `tokensToMegaTokens` outputs very small numbers.
   - **Mitigation**:
     - Use libraries like `decimal.js` for precise arithmetic.
     - Ensure that all intermediate results are rounded or truncated appropriately.

3. **Usage Calculators (e.g., `TokenBasedCostCalculator`)**:
   - Similar risks and mitigations apply as in the points above. These calculators involve direct floating-point operations.

---

### General Observations:
- Floating-point operations are distributed across multiple files (e.g., `pricing.ts`, `calculator.ts`, `model-prices.ts`), primarily in cost and usage calculations.
- These calculations involve multiplication and division, which are prone to precision issues in JavaScript/TypeScript.

### Recommendations:
1. **Validation and Testing**:
   - Add unit tests for edge cases such as:
     - Very large token counts.
     - Very small token counts (e.g., `1` or `10`).
     - Rounding consistency in cost calculations.
   - Ensure test coverage for scenarios where operations might lose precision.

2. **Precision Control**:
   - Use a fixed precision for results where possible (e.g., `toFixed()` or `Number.toPrecision()`).
   - Consider using libraries like `big.js` or `decimal.js` for calculations requiring high precision.

3. **Error Handling**:
   - Validate and sanitize inputs to prevent unexpected behavior (e.g., negative or zero token counts).

---

### Further Action:
The pull request includes several modified and new files. To ensure a comprehensive review, I recommend focusing on:
- `packages/language-model/src/costs/pricing.ts`
- `packages/language-model/src/costs/calculator.ts`
- `packages/language-model/src/costs/model-prices.ts`

Let me know if you'd like detailed feedback on any specific section!

File: packages/language-model/src/costs/pricing.ts [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+	constructor(
+		private readonly pricing: Record<string, { prices: ModelPrice[] }>,
+	) {}
+
+	calculateCost(modelId: string, usage: TokenUsage) {
+		const pricing = this.pricing[modelId]?.prices[0]?.price;
+		if (!pricing) return null;
+
+		const inputCost =
+			tokensToMegaTokens(usage.input) * pricing.input.costPerMegaToken;
```
Comment (shige) [inline packages/language-model/src/costs/pricing.ts]: **Example of a floating-point problem**
```js
// usage.input = 300
// pricing.input.costPerMegaToken = 0.1

// tokensToMegaTokens(usage.input)
300 / 1_000_000
> 0.0003

// tokensToMegaTokens(usage.input) * pricing.input.costPerMegaToken
> 0.0003 * 0.1
0.000029999999999999997
```

Comment (shige) [reply packages/language-model/src/costs/pricing.ts]: **Example of converting to an integer value and then calculating: (without using libraries such as big.js)**

```js
> ( 300 * (0.1 * 100) ) / (100 * 1_000_000)
0.00003

> ( 300 * (0.1 * 1_000_000) ) / (1_000_000 * 1_000_000)
0.00003
```

**Review by shige — CHANGES_REQUESTED** [review]
There seems to be an issue with Floating-Point Calculations, please fix it!

File: packages/language-model/src/costs/pricing.ts (hunk 9 above)
Comment (rindrics) [reply packages/language-model/src/costs/pricing.ts]: fixed: https://github.com/giselles-ai/giselle/pull/834/commits/9f1a818889819b664cb2ccda2ebb15397518ef15

File: packages/language-model/src/costs/pricing.ts (hunk 7 above)
Comment (rindrics) [reply packages/language-model/src/costs/pricing.ts]: Thank you, I added tests to `calculateCost()` by commits after https://github.com/giselles-ai/giselle/pull/834/commits/13eb092d6af43862c8d1d3be7b448205604579e3

File: packages/language-model/src/costs/pricing.ts (hunk 8 above)
Comment (rindrics) [reply packages/language-model/src/costs/pricing.ts:5]: You're right.
We need another PR focusing on cost calculation logic.

However, I think we can merge this PR because:
- the calculation errors were avoided using integer-conversion method https://github.com/giselles-ai/giselle/pull/834/commits/9f1a818889819b664cb2ccda2ebb15397518ef15
- we don't bill users by model-usage at this point
  - we will do this after implementing robust & clear cost-calculation logic

@shige

File: packages/language-model/src/costs/calculator.ts [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+	const outputCostPerMegaTokenInCents =
+		pricing.output.costPerMegaToken * dollarToCent * tokensPerMegaToken;
+
+	// Calculate costs in cents using integer arithmetic
+	const inputCostInCents = usage.promptTokens * inputCostPerMegaTokenInCents;
+	const outputCostInCents =
+		usage.completionTokens * outputCostPerMegaTokenInCents;
+	const totalCostInCents = inputCostInCents + outputCostInCents;
+
+	// Convert back to dollars by dividing by 1,000,000 (mega tokens) and 100 (cents)
```
[the remark this answers is not in the document]
Comment (rindrics) [reply packages/language-model/src/costs/calculator.ts:59]: I'll remain the comment as it is because I think the suggested version is too much

**Comment by rindrics** [comment]
📝 we should group trace by workspace ID not generation ID

File: packages/language-model/src/costs/pricing.ts (hunk 9 above)
Comment (shige) [reply packages/language-model/src/costs/pricing.ts]: @Rindrics I confirmed that calculateTokenCost has been fixed, but calculateCost does not seem to have been fixed yet.
https://github.com/giselles-ai/giselle/pull/834/files#diff-1b0de3799947b0b3eabd115b38249c31c721490baa840d97eb68051cbb6c5763R47

File: packages/language-model/src/costs/calculator.ts [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+export class DefaultCostCalculator implements CostCalculator {
+	constructor(private provider: string) {}
+
+	async calculate(model: string, usage: ModelUsage): Promise<CostResult> {
+		console.log(`No cost calculator found for ${this.provider}`);
+		return { input: 0, output: 0, total: 0 };
+	}
+}
+
+export function calculateTokenCost(
```
Comment (shige) [inline packages/language-model/src/costs/calculator.ts]: IMO:

I think the codes and comments are difficult to understand, so I would like you to refactor it, even if it's a separate issue or PR.

**Comment by shige** [comment]
@toyamarinyon @satococoa 
Since this will be the base implementation for Giselle's fee calculation, I would like to ask all developers to review this PR!

File: packages/language-model/src/costs/pricing.ts (hunk 9 above)
Comment (rindrics) [reply packages/language-model/src/costs/pricing.ts]: sorry, I deleted it by ae0dce90 because it was not referenced @shige

File: packages/giselle-engine/src/core/generations/generate-text.ts [hunk 12]
```diff
@@ -78,6 +85,23 @@ export async function generateText(args: {
...
+	const generation = span.generation({
+		name: "ai.streamText.doStream",
+		model: operationNode.content.llm.id,
+		modelParameters: operationNode.content.llm.configurations,
+		startTime: new Date(runningGeneration.createdAt),
+		completionStartTime: new Date(runningGeneration.startedAt),
+	});
+
```
Comment (toyamarinyon) [inline packages/giselle-engine/src/core/generations/generate-text.ts]: Why do you need to directly operate the Langfuse instance? I would like to know the reason why it cannot be done within the framework of AI SDK Telemetry.
https://ai-sdk.dev/docs/ai-sdk-core/telemetry

Also, if it cannot be done within the framework of AI SDK Telemetry, is it not possible to create a Custom Tracer to address this?
https://ai-sdk.dev/docs/ai-sdk-core/telemetry#custom-tracer

File: packages/giselle-engine/src/core/generations/generate-text.ts [hunk 13]
```diff
@@ -78,6 +85,23 @@ export async function generateText(args: {
 		startedAt: Date.now(),
 	} satisfies RunningGeneration;
 
+	const trace = langfuse.trace({
+		userId: String(args.telemetry?.metadata?.userId),
+		name: "ai.streamText",
```
Comment (satococoa) [inline packages/giselle-engine/src/core/generations/generate-text.ts]: Is this a recommended or default naming by Langfuse?

File: packages/giselle-engine/src/core/generations/generate-text.ts [hunk 14]
```diff
@@ -78,6 +85,23 @@ export async function generateText(args: {
...
+		metadata: args.telemetry?.metadata,
+	});
+	const span = trace.span({
+		name: "ai.streamText",
+		startTime: new Date(runningGeneration.queuedAt),
```
Comment (satococoa) [inline packages/giselle-engine/src/core/generations/generate-text.ts]: I wonder which is better, runningGeneration.queuedAt or runningGeneration.startedAt?
I'd like to get your thoughts on using runningGeneration.queuedAt versus runningGeneration.startedAt.

File: packages/giselle-engine/src/core/generations/generate-text.ts [hunk 15]
```diff
@@ -396,6 +434,31 @@ export async function generateText(args: {
...
+			updateTelemetry(span, { messages }, event.text);
+			updateTelemetry(generation, { messages }, event.text);
+			const tokenUsage = event.usage;
+
+			let costInfo = null;
+			const provider = operationNode.content.llm.provider;
+			const modelId = operationNode.content.llm.id;
+
+			if (tokenUsage) {
+				const { calculateCost } = await import("@giselle-sdk/language-model");
```
Comment (satococoa) [inline packages/giselle-engine/src/core/generations/generate-text.ts]: Can we use static(normal) imports instead of dynamic imports in this module?

File: packages/giselle-engine/src/core/generations/telemetry.ts [hunk 16]
```diff
@@ -1,16 +1,35 @@
...
+
+type TelemetryTag =
+	| BaseFunctionalityTag
+	| ProviderNameTag
+	| ProviderOptionTag
+	| ModelNameTag;
+
+let langfuseInstance: Langfuse | null = null;
+
+export function getLangfuseInstance(): Langfuse {
```
Comment (satococoa) [inline packages/giselle-engine/src/core/generations/telemetry.ts]: Do we need to implement this as a Singleton? If new Langfuse() is a heavy operation, this approach makes sense. But if it's not resource-intensive, creating new instances would prevent unintended state sharing.

File: packages/giselle-engine/src/core/generations/generate-text.ts [hunk 17]
```diff
@@ -78,6 +85,23 @@ export async function generateText(args: {
 		startedAt: Date.now(),
 	} satisfies RunningGeneration;
 
+	const trace = langfuse.trace({
+		userId: String(args.telemetry?.metadata?.userId),
```
Comment (satococoa) [inline packages/giselle-engine/src/core/generations/generate-text.ts]: Since teams are our billing entities, I'm wondering why we need userId in the telemetry data. Wouldn't teamId alone be sufficient for billing purposes?

File: packages/giselle-engine/src/core/generations/generate-text.ts (hunk 12 above)
Comment (rindrics) [reply packages/giselle-engine/src/core/generations/generate-text.ts]: > Why do you need to directly operate the Langfuse instance?

we need Langfuse instance because [`experimental_telemetry` of `streamText()` does not have attribute`cost`](https://ai-sdk.dev/docs/reference/ai-sdk-core/stream-text#experimental_telemetry.telemetry-settings).

> is it not possible to create a Custom Tracer to address this?
> https://ai-sdk.dev/docs/ai-sdk-core/telemetry#custom-tracer

yes, I think we can use our own custom tracer (not `tracerProvider.getTracer('ai')` shown in the reference) in the future.
Let me try this in another PR.

File: packages/language-model/src/costs/calculator.ts [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+		pricing.output.costPerMegaToken * dollarToCent * tokensPerMegaToken;
+
+	// Calculate costs in cents using integer arithmetic
+	const inputCostInCents = usage.promptTokens * inputCostPerMegaTokenInCents;
+	const outputCostInCents =
+		usage.completionTokens * outputCostPerMegaTokenInCents;
+	const totalCostInCents = inputCostInCents + outputCostInCents;
+
+	// Convert back to dollars by dividing by 1,000,000 (mega tokens) and 100 (cents)
+	const divisor = tokensPerMegaToken * dollarToCent * tokensPerMegaToken;
```
Comment (satococoa) [inline packages/language-model/src/costs/calculator.ts:60]: The `Calculator` should ideally return precise values.

Returning a floating-point result here may introduce precision errors, which is an anti-pattern for handling money in JavaScript.

I recommend performing all calculations in integers and only converting to floating-point for display.

File: packages/giselle-engine/src/core/generations/generate-text.ts (hunk 12 above)
Comment (toyamarinyon) [reply packages/giselle-engine/src/core/generations/generate-text.ts]: Generation is the core of the core function in Giselle. Although you might think that telemetry could also be implemented in the same detailed way, I believe that telemetry should have a thin implementation that doesn't make its presence felt. (Alternatively, there could be a design that wraps it as a template function like in the AI SDK.) Looking forward to the implementation.

**Review by toyamarinyon — APPROVED** [review]

File: packages/language-model/src/costs/model-prices.ts [hunk 19]
```diff
@@ -0,0 +1,178 @@
+import type { ModelPrice } from "./pricing";
+
+export type ModelPriceTable = Record<string, { prices: ModelPrice[] }>;
+
+const validPriceCache = new Map<string, ModelPrice>();
```
Comment (satococoa) [inline packages/language-model/src/costs/model-prices.ts]: Is implementing a cache necessary here?
Directly retrieving the current valid price seems adequate and less error-prone.

File: packages/language-model/src/costs/model-prices.ts [hunk 20]
```diff
@@ -0,0 +1,178 @@
...
+
+let lastCacheResetDate = new Date();
+
+// Check and reset cache if it's a new UTC day
+function checkAndResetCacheIfNeeded() {
```
Comment (satococoa) [inline packages/language-model/src/costs/model-prices.ts]: We would have to consider time precision (hour, minutes, seconds) if we implement a cache mechanism.

**Review by satococoa — CHANGES_REQUESTED** [review]
I've added some comments!

File: packages/giselle-engine/src/core/generations/generate-text.ts (hunk 12 above)
Comment (rindrics) [reply packages/giselle-engine/src/core/generations/generate-text.ts]: > I believe that telemetry should have a thin implementation that doesn't make its presence felt

I agree, it's always true!
I'll refactor them into the form they should be.

File: packages/giselle-engine/src/core/generations/generate-text.ts (hunk 13 above)
Comment (rindrics) [reply packages/giselle-engine/src/core/generations/generate-text.ts]: no, we should reconsider their (trace, span, generation) naming convention in the future.

File: packages/giselle-engine/src/core/generations/generate-text.ts (hunk 14 above)
Comment (rindrics) [reply packages/giselle-engine/src/core/generations/generate-text.ts]: Thanks, I set `runningGeneration.startedAt` to the start time of `generation` (not `span`, focused here).

I think my configuration is reasonable, but there is a room to discuss this if we have to improve speed performance of generation node.
Please let me leave this configuration as it is in this PR.

File: packages/giselle-engine/src/core/generations/generate-text.ts (hunk 15 above)
Comment (rindrics) [reply packages/giselle-engine/src/core/generations/generate-text.ts]: Thanks, I simplified the code using static import by 3f1994c4

File: packages/giselle-engine/src/core/generations/telemetry.ts (hunk 16 above)
Comment (rindrics) [reply packages/giselle-engine/src/core/generations/telemetry.ts]: I thought sharing Langfuse instance itself as singleton is OK because every generation creates new trace instance by

```node
const trace = langfuse.trace();
```

But after reading your comment I feel I should reconsider about this

because I couldn't find articles that suggests to create Langfuse instance as a Singleton 🤔

File: packages/giselle-engine/src/core/generations/generate-text.ts (hunk 15 above)
Comment (toyamarinyon) [reply packages/giselle-engine/src/core/generations/generate-text.ts]: I wonder if it wasn't using "await import" because it was necessary? Why was it initially implemented with "await import"?

I'm curious about the background since I don't think it would have ended up this way with a normal implementation.

File: packages/giselle-engine/src/core/generations/telemetry.ts (hunk 16 above)
Comment (rindrics) [reply packages/giselle-engine/src/core/generations/telemetry.ts]: I will revert edfa68a1e46ad24a0f08258cfa2816bee2b76b73

File: packages/giselle-engine/src/core/generations/generate-text.ts (hunk 17 above)
Comment (rindrics) [reply packages/giselle-engine/src/core/generations/generate-text.ts]: > Wouldn't teamId alone be sufficient for billing purposes

Yes, it is sufficient for the billing purposes.

This is for an analytical purpose: Langfuse provides functionality to track usage of single user
So let me leave the code as it is 👍 

The user in the screenshot below is me on the development environment:
image

File: packages/language-model/src/costs/calculator.ts (hunk 18 above)
Comment (rindrics) [reply packages/language-model/src/costs/calculator.ts:60]: Thanks,

> I recommend performing all calculations in integers and only converting to floating-point for display.

This is great if we could this "always", but in this situation we can't.
Because the cost of single generation will be in float if we adopt cents for our calculation unit:

For example,
- cost of "mega" token is 200 cents (2.0 dollar) in gpt-4.1
  - so if the tokens consumed by a single generation is less than 5000, the calculated cost in cents will be smaller than 1 cents
  - > if the tokens consumed by a single generation is less than 5000
    - and this happens so often

Please let me leave the code here as it is because we already avoided the calculation errors were avoided using integer-conversion method https://github.com/giselles-ai/giselle/commit/9f1a818889819b664cb2ccda2ebb15397518ef15

File: packages/language-model/src/costs/model-prices.ts (hunk 19 above)
Comment (rindrics) [reply packages/language-model/src/costs/model-prices.ts]: Yes, I'll consider the speed-related performance solution "after" speed-related problem occurred.
We should focus on accuracy first.

Comment (rindrics) [reply packages/language-model/src/costs/model-prices.ts]: I'll revert ea8d349da81619663831e5b4c6ad04d7aca4269e

File: packages/giselle-engine/src/core/generations/generate-text.ts (hunk 17 above)
Comment (satococoa) [reply packages/giselle-engine/src/core/generations/generate-text.ts]: got it! 👍

File: packages/giselle-engine/src/core/generations/generate-text.ts (hunk 15 above)
Comment (rindrics) [reply packages/giselle-engine/src/core/generations/generate-text.ts]: Sorry, it was not by necessity.

The first version simply loaded the function only when `tokenUsage` was non-null,
but because the function is not heavy, moreover, dynamic imports are basically not used in this repository, I updated this line to use normal import. @toyamarinyon

File: packages/language-model/src/costs/model-prices.ts (hunk 19 above)
Comment (rindrics) [reply packages/language-model/src/costs/model-prices.ts]: reverted by a97de5ad

File: packages/language-model/src/costs/model-prices.ts (hunk 20 above)
Comment (rindrics) [reply packages/language-model/src/costs/model-prices.ts]: Thank you.
This comment became outdated because I quit caching by a97de5ad

File: packages/giselle-engine/src/core/generations/telemetry.ts (hunk 16 above)
Comment (rindrics) [reply packages/giselle-engine/src/core/generations/telemetry.ts]: reverted by c62d7e5d

**Comment by rindrics** [comment]
Thank you, I replied to your each comment.
Please let me know if you still have any questions @satococoa

File: packages/language-model/src/costs/calculator.ts (hunk 18 above)
Comment (satococoa) [reply packages/language-model/src/costs/calculator.ts:60]: > the cost of single generation will be in float

I see.

Since we offer monthly billing, the amount of model usage fees is a sum of these 'costs of single generation'.
For display purposes of a single generation, this float value is sufficient and rounding is fine.
However, we cannot use these rounded values to calculate the sum of monthly model usage accurately.

Is this logic and return value only used for display purposes, or do you plan to use this value to calculate monthly fees as well?

----

For example, here are my ideas about how to calculate "monthly fees":

## Plan A: Record calculated price in sufficient precision integers

Instead of using floating-point numbers, we can decide on an appropriate level of precision (e.g., millicents, microcents) and store all prices as integers.

For example:
- $1.25 would be stored as 1,250,000 microcents
- $0.00123 would be stored as 1,230 microcents

This approach:
- Avoids floating-point errors completely
- Allows exact summing of values
- Requires only one division operation when displaying the final amount
- Is compatible with most database systems

## Plan B: Record token value

In this plan, rounding is only occur only once when calculating monthly fees.

Comment (satococoa) [reply packages/language-model/src/costs/calculator.ts:60]: Since calculation monthly fees is out of this pull request's scope, I want to know here is only this: 🙏 

> Is this logic and return value only used for display purposes, or do you plan to use this value to calculate monthly fees as well?

Comment (rindrics) [reply packages/language-model/src/costs/calculator.ts:60]: Ah, I finally understand what you meant 🙇 

Item | Option A (Store cost in millicents) | Option B (Store token count only)
-- | -- | --
Implementation Simplicity | Calculate and store cost per request | Store only token count, calculate later
Flexibility (Rate Changes) | Low (hard to adjust past records) | High (can recalculate past costs easily)
Accuracy (Rounding Errors) | Very high (integer-based) | Requires care to avoid rounding errors
Billing Aggregation Logic | Simple (just sum stored amounts) | More complex (match token usage with rates)
Storage Efficiency | Lower (store cost every time) | Higher (store only token usage)

I suggest `Option B`.
The drawback of the option, "Requires care to avoid rounding errors", won't be a problem if we clarify the rounding logic (on the monthly batch) by document or FAQ

What do you think? @satococoa cc: @shige

Comment (rindrics) [reply packages/language-model/src/costs/calculator.ts:60]: > Is this logic and return value only used for display purposes, or do you plan to use this value to calculate monthly fees as well?

Firstly I wrote with the intention of fee calculations, but now I realized that at least with current implementation, the calculated value should be used for "display only"! @satococoa

Comment (satococoa) [reply packages/language-model/src/costs/calculator.ts:60]: > I suggest Option B.

Good 👍 
And I understand this method is intended to be used for "display only".

File: packages/giselle-engine/src/core/generations/telemetry.ts [hunk 21]
```diff
@@ -1,48 +1,160 @@
...
+			unit,
+		},
+		startTime: new Date(runningGeneration.createdAt),
+		completionStartTime: new Date(runningGeneration.startedAt),
+		metadata: settings?.metadata,
+		output,
+		endTime: new Date(completedGeneration.completedAt),
+	});
+
+	return { langfuse, trace, span, generation };
```
Comment (rindrics) [inline packages/giselle-engine/src/core/generations/telemetry.ts:163]: need to return not only Langfuse object, but also other child tracer objects to get telemetry data correctly sent to Langfuse

**Comment by rindrics** [comment]
From your last review, I updated the code in the points below:
- explain that cost calculation function is for "display-only" f3509425
- minimize exposure of telemetry codes in `generate-text.ts` b26e169d

would you review the code again? @satococoa

File: packages/giselle-engine/src/core/generations/generate-text.ts [hunk 22]
```diff
@@ -469,6 +484,37 @@ export async function generateText(args: {
 				generation: completedGeneration,
 				onConsumeAgentTime: args.context.onConsumeAgentTime,
 			});
+
+			console.log("before---------------");
```
Comment (satococoa) [inline packages/giselle-engine/src/core/generations/generate-text.ts]:
```suggestion
```

File: packages/giselle-engine/src/core/generations/generate-text.ts [hunk 23]
```diff
@@ -469,6 +484,37 @@ export async function generateText(args: {
...
+					outputCost: costInfo?.outputCost ?? 0,
+					totalCost: costInfo?.totalCost ?? 0,
+				},
+				completedGeneration,
+				spanName: "ai.streamText",
+				generationName: "ai.streamText.doStream",
+				unit: "TOKENS",
+				settings: args.telemetry,
+			});
+			console.log("after---------------");
```
Comment (satococoa) [inline packages/giselle-engine/src/core/generations/generate-text.ts]:
```suggestion
```

File: packages/giselle-engine/src/core/generations/telemetry.ts [hunk 24]
```diff
@@ -59,3 +76,93 @@ export function generateTelemetryTags(args: {
...
+	settings?: TelemetrySettings;
+}): {
+	langfuse: Langfuse;
+	trace: ReturnType<Langfuse["trace"]>;
+	span: ReturnType<ReturnType<Langfuse["trace"]>["span"]>;
+	generation: ReturnType<
+		ReturnType<ReturnType<Langfuse["trace"]>["span"]>["generation"]
+	>;
+} {
+	console.log("inside======");
```
Comment (satococoa) [inline packages/giselle-engine/src/core/generations/telemetry.ts]: Are leaving these `console.log()`s OK for us? or should remove? cc @shige

File: packages/giselle-engine/src/core/generations/telemetry.ts (hunk 21 above)
Comment (satococoa) [reply packages/giselle-engine/src/core/generations/telemetry.ts:163]: It appears that the returned objects in `generate-text.ts` are not being utilized. 
Should we still return these objects, or would it be better to refactor this code?

File: packages/language-model/src/costs/index.ts [hunk 25]
```diff
@@ -0,0 +1,40 @@
...
+	switch (provider) {
+		case "openai":
+			return new OpenAICostCalculator();
+		default:
+			console.log(`Unimplemented provider: ${provider}`);
+			return new DefaultCostCalculator(provider);
+	}
+}
+
+export async function calculateCost(
```
Comment (satococoa) [inline packages/language-model/src/costs/index.ts]: I think this is a bit misleading.

Since the calculation method used in this logic is named `calculateTokenCostForDisplay`, hiding this intention could be problematic.

If it's acceptable that the price value sent to Langfuse is a display price, then the metric name and calculation method should be named something like `displayPrice`.

**Review by satococoa — CHANGES_REQUESTED** [review]
Thank you!
I've added some additional comments 💬

File: packages/giselle-engine/src/core/generations/telemetry.ts (hunk 24 above)
Comment (rindrics) [reply packages/giselle-engine/src/core/generations/telemetry.ts]: sorry, I removed it by e198bc2a 🙇

File: packages/giselle-engine/src/core/generations/generate-text.ts (hunk 23 above)
Comment (rindrics) [reply packages/giselle-engine/src/core/generations/generate-text.ts]: fixed: e198bc2a

File: packages/giselle-engine/src/core/generations/generate-text.ts (hunk 22 above)
Comment (rindrics) [reply packages/giselle-engine/src/core/generations/generate-text.ts]: fixed: e198bc2a

File: packages/giselle-engine/src/core/generations/telemetry.ts (hunk 21 above)
Comment (rindrics) [reply packages/giselle-engine/src/core/generations/telemetry.ts:163]: > Should we still return these objects

yes, need to return to get telemetry sent to Langfuse!

> or would it be better to refactor this code?

yes, I renamed these objects with adding prefix `_`: a18b17e0

File: packages/language-model/src/costs/index.ts (hunk 25 above)
Comment (rindrics) [reply packages/language-model/src/costs/index.ts]: Thank you.
I renamed objects that may mislead the their intention: d602f5db

File: packages/giselle-engine/src/core/generations/generate-text.ts [hunk 26]
```diff
@@ -469,6 +484,41 @@ export async function generateText(args: {
...
 				onConsumeAgentTime: args.context.onConsumeAgentTime,
 			});
+
+			// necessary to send telemetry but not explicitly used
+			const {
+				langfuse: _langfuse,
+				trace: _trace,
+				span: _span,
+				generation: _generation,
+			} = createLangfuseTracer({
```
Comment (satococoa) [inline packages/giselle-engine/src/core/generations/generate-text.ts]: Nit: If we need to keep these objects for telemetry, maybe we could write the code like this instead:

```suggestion
			const _langfuseTracer = createLangfuseTracer({
```

**Review by satococoa — APPROVED** [review]
Thank you!

I think the naming and structure of price calculation functions have room for refinement, but it would be better to restructure them when we implement monthly pricing.

File: packages/giselle-engine/src/core/generations/generate-text.ts (hunk 26 above)
Comment (rindrics) [reply packages/giselle-engine/src/core/generations/generate-text.ts]: Thank you, I confirmed that the telemetries arrived after applying the change suggestion above! ✅

**Comment by rindrics** [comment]
We decided to implement cost calculation for "billing" in https://github.com/giselles-ai/giselle/issues/865.
Now we can merge this PR because it calculates cost for displaying preliminary value @shige

**Comment by rindrics** [comment]
Thank you for your reviews!
